recreate.go

```diff
@@ -0,0 +1,437 @@
+// +build !cassandra
+// Copyright (C) 2017 ScyllaDB
+
+package gocql
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+	"text/template"
+)
+
+func (km *KeyspaceMetadata) Recreate() (string, error) {
+	sb := &strings.Builder{}
+
+	if err := km.recreateKeyspace(sb); err != nil {
+		return "", err
+	}
+
+	sortedTypes := km.typesSortedTopologically()
+	for _, tm := range sortedTypes {
+		if err := km.recreateUserTypes(sb, tm); err != nil {
+			return "", err
+		}
+	}
+
+	for _, tm := range km.Tables {
+
+		if err := km.recreateTable(sb, km.Name, tm); err != nil {
+			return "", err
+		}
+	}
+
+	for _, im := range km.Indexes {
+		if err := km.recreateIndex(sb, im); err != nil {
+			return "", err
+		}
+	}
+
+	for _, fm := range km.Functions {
+		if err := km.recreateFunction(sb, km.Name, fm); err != nil {
+			return "", err
+		}
+	}
+
+	for _, am := range km.Aggregates {
+		if err := km.recreateAggregate(sb, am); err != nil {
+			return "", err
+		}
+	}
+
+	for _, vm := range km.Views {
+		if err := km.recreateView(sb, vm); err != nil {
+			return "", err
+		}
+	}
+
+	return sb.String(), nil
+}
+
+func (km *KeyspaceMetadata) typesSortedTopologically() []*TypeMetadata {
+	sortedTypes := make([]*TypeMetadata, 0, len(km.Types))
+	for _, tm := range km.Types {
+		sortedTypes = append(sortedTypes, tm)
+	}
+	sort.Slice(sortedTypes, func(i, j int) bool {
+		for _, ft := range sortedTypes[j].FieldTypes {
+			if strings.Contains(ft, sortedTypes[i].Name) {
+				return true
+			}
+		}
+		return false
+	})
+	return sortedTypes
+}
+
+func (km *KeyspaceMetadata) recreateTable(w io.Writer, kn string, tm *TableMetadata) error {
+	t := template.Must(template.New("table").
+		Funcs(map[string]interface{}{
+			"escape":                  km.escape,
+			"recreateTableColumns":    km.recreateTableColumns,
+			"recreateTableProperties": km.recreateTableProperties,
+		}).
+		Parse(`
+CREATE TABLE {{ .KeyspaceName }}.{{ .Tm.Name }} (
+  {{ recreateTableColumns .Tm }}
+) WITH {{ recreateTableProperties .Tm.ClusteringColumns .Tm.Options .Tm.Flags }};
+`))
+	if err := t.Execute(w, map[string]interface{}{
+		"Tm":           tm,
+		"KeyspaceName": kn,
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (km *KeyspaceMetadata) recreateTableColumns(tm *TableMetadata) string {
```

Comment:
Strange API you create a new builder to get a stirring to put in a template to write to another builder.
It sort of does not make sense.
Not a performance problem here.

Reply:
These helpers are shared across different entities, this particular one is used in `table` and `view` entity. 
It's not possible to pass same string builder to helper function and main function using template. 
They also exists because it was easier and clearer to implement logic in Go instead of in template.